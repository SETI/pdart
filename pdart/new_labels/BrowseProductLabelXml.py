"""A document template to create a label for a raw browse product."""
from pdart.xml.Pds4Version import *
from pdart.xml.Templates import *

make_label = interpret_document_template(
    """<?xml version="1.0" encoding="utf-8"?>
<?xml-model href="http://pds.nasa.gov/pds4/pds/v1/PDS4_PDS_%s.sch"
            schematypens="http://purl.oclc.org/dsdl/schematron"?>
<Product_Browse
   xmlns="http://pds.nasa.gov/pds4/pds/v1"
   xmlns:pds="http://pds.nasa.gov/pds4/pds/v1"
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://pds.nasa.gov/pds4/pds/v1/PDS4_PDS_%s.xsd">
  <Identification_Area>
    <logical_identifier><NODE name="browse_lid" /></logical_identifier>
    <version_id>0.1</version_id>
    <title>This product contains a browse image of a <NODE name="suffix" /> \
image obtained the HST Observing Program <NODE name="proposal_id" />\
.</title>
    <information_model_version>%s</information_model_version>
    <product_class>Product_Browse</product_class>
    <Modification_History>
      <Modification_Detail>
        <modification_date>2016-04-20</modification_date>
        <version_id>0.1</version_id>
        <description>PDS4 version-in-development of the product</description>
      </Modification_Detail>
    </Modification_History>
  </Identification_Area>
  <Reference_List>
    <Internal_Reference>
      <lid_reference><NODE name="data_lid" /></lid_reference>
      <reference_type>browse_to_data</reference_type>
    </Internal_Reference>
  </Reference_List>
  <File_Area_Browse>
    <File>
      <file_name><NODE name="browse_file_name" /></file_name>
    </File>
    <Encoded_Image>
      <offset unit="byte">0</offset>
      <object_length unit="byte"><NODE name="object_length" /></object_length>
      <encoding_standard_id>JPEG</encoding_standard_id>
    </Encoded_Image>
  </File_Area_Browse>
</Product_Browse>""" %
    (PDS4_SHORT_VERSION, PDS4_SHORT_VERSION, INFORMATION_MODEL_VERSION))
# type: DocTemplate
"""
An interpreted document template to create a label for a RAW browse product.
"""
