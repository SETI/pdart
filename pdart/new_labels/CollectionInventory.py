"""
Functionality to build a collection inventory using a SQLite database.
"""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


def make_collection_inventory(bundle_db, collection_lidvid):
    # type: (BundleDB, str) -> unicode
    products = bundle_db.get_collection_products(
        collection_lidvid)
    inventory_lines = [u'P,%s\r\n' % product.lidvid
                       for product in products]
    return ''.join(inventory_lines)
