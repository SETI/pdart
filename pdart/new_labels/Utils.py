"""
Utility functions.
"""
from typing import TYPE_CHECKING
import unittest
from fs.path import dirname, join
import io
from os.path import isfile

from pdart.pds4.LIDVID import LIDVID

if TYPE_CHECKING:
    from typing import Union


def lidvid_to_lid(lidvid):
    # type: (str) -> str
    """Return the LID of the LIDVID."""
    return str(LIDVID(lidvid).lid())


def lidvid_to_vid(lidvid):
    # type: (str) -> str
    """Return the VID of the LIDVID."""
    return str(LIDVID(lidvid).vid())


def _golden_filepath(basename):
    # type: (unicode) -> unicode
    """Return the path to a golden file with the given basename."""
    return join(dirname(__file__), basename)


def _golden_file_contents(basename, is_inventory):
    # type: (unicode, bool) -> unicode
    """
    Return the contents as a Unicode string of the golden file with
    the given basename.
    """
    if is_inventory:
        with open(_golden_filepath(basename), 'rb') as f:
            return f.read()
    else:
        with io.open(_golden_filepath(basename)) as f:
            return f.read()

def assert_golden_file_equal(testcase, basename,
                             calculated_contents, is_inventory=False):
    # type: (unittest.TestCase, unicode, Union[str, unicode], bool) -> None
    """
    Finds the golden file and compares its contents with the given string.
    Raises an exception via the unittest.TestCase argument if they are
    unequal.  If the file does not yet exist, it writes the given string to
    the file.  Inventories must be written with CR/NL line-termination, so
    a parameter for that is provided.
    """

    filepath = _golden_filepath(basename)
    if isfile(filepath):
        testcase.assertEqual(_golden_file_contents(filepath, is_inventory),
                             calculated_contents)
    else:
        if is_inventory:
            assert calculated_contents[-2:] == '\r\n', 'inventory input is wrong'
            with open(filepath, 'wb') as f:
                f.write(calculated_contents)
            with open(filepath, 'rb') as f:
                roundtrip = f.read()
                assert roundtrip[-2:] == '\r\n', 'inventory roundtripped is wrong'
        else:
            with io.open(filepath, 'w') as f:
                f.write(unicode(calculated_contents))
        testcase.fail('Golden file %r did not exist but it was created.' %
                      basename)

def path_to_testfile(basename):
    # type: (unicode) -> unicode
    """Return the path to files needed for testing."""
    return join(dirname(__file__), 'testfiles', basename)
