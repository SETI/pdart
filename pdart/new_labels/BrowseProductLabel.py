"""
Functionality to create a label for a browse product containing browse
images.
"""
from typing import TYPE_CHECKING

from pdart.new_labels.BrowseProductLabelXml import make_label
from pdart.new_labels.Utils import lidvid_to_lid, lidvid_to_vid
from pdart.xml.Pretty import pretty_and_verify

if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


def make_browse_product_label(bundle_db,
                              browse_product_lidvid,
                              browse_file_basename,
                              verify):
    # type: (BundleDB, str, str, unicode, bool) -> unicode
    """
    Create the label text for the browse product having the given
    LIDVID using the bundle database.  If verify is True, verify the
    label against its XML and Schematron schemas.  Raise an exception
    if either fails.
    """
    fits_product_lidvid = bundle_db.get_product(
        browse_product_lidvid).fits_product_lidvid
    browse_file = bundle_db.get_file(browse_file_basename,
                                     browse_product_lidvid)

    browse_product = bundle_db.get_product(browse_product_lidvid)

    browse_collection_lidvid = browse_product.collection_lidvid
    browse_collection = bundle_db.get_collection(browse_collection_lidvid)

    bundle = bundle_db.get_bundle()

    label = make_label({
        'proposal_id': bundle.proposal_id,
        'suffix': browse_collection.suffix,
        'browse_lid': lidvid_to_lid(browse_product_lidvid),
        'browse_vid': lidvid_to_vid(browse_product_lidvid),
        'data_lidvid': fits_product_lidvid,
        'browse_file_name': browse_file_basename,
        'object_length': browse_file.byte_size
    }).toxml()

    return pretty_and_verify(label, verify)
