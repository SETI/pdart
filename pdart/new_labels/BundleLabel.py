"""Functionality to build a bundle label using a SQLite database."""
from typing import TYPE_CHECKING

from pdart.new_labels.BundleLabelXml \
    import make_bundle_entry_member, make_label
from pdart.new_labels.CitationInformation \
    import make_citation_information
from pdart.new_labels.Utils import lidvid_to_lid, lidvid_to_vid
from pdart.xml.Pretty import pretty_and_verify
from pdart.xml.Templates import combine_nodes_into_fragment

if TYPE_CHECKING:
    from Citation_Information import Citation_Information
    from pdart.new_db.BundleDB import BundleDB


def make_bundle_label(bundle_db, info, verify):
    # type: (BundleDB, Citation_Information, bool) -> unicode
    """
    Create the label text for the bundle in the bundle database using
    the database connection.  If verify is True, verify the label
    against its XML and Schematron schemas.  Raise an exception if
    either fails.
    """
    bundle = bundle_db.get_bundle()
    bundle_lid = lidvid_to_lid(bundle.lidvid)  # Only used in placeholder
    proposal_id = bundle.proposal_id
    reduced_collections = [
        make_bundle_entry_member({'collection_lidvid': collection.lidvid})
        for collection
        in bundle_db.get_bundle_collections(bundle.lidvid)]

    label = make_label({
        'bundle_lid': lidvid_to_lid(bundle.lidvid),
        'bundle_vid': lidvid_to_vid(bundle.lidvid),
        'proposal_id': str(proposal_id),
        'Citation_Information': make_citation_information(info),
        'Bundle_Member_Entries': combine_nodes_into_fragment(
            reduced_collections)
    }).toxml()

    return pretty_and_verify(label, verify)
