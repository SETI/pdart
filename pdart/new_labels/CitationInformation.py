"""
Functionality to create a ``<Citation_Information />`` XML element.
"""
from typing import TYPE_CHECKING

from Citation_Information import Citation_Information
from pdart.new_labels.Placeholders import known_placeholder, placeholder_year
from pdart.xml.Templates import interpret_template

if TYPE_CHECKING:
    from pdart.xml.Templates import NodeBuilder, NodeBuilderTemplate

_citation_information = interpret_template(
    """<Citation_Information>
<publication_year><NODE name="publication_year" /></publication_year>
<description><NODE name="description" /></description>
</Citation_Information>""")  # type: NodeBuilderTemplate
"""
An interpreted template to create a ``<Citation_Information
/>`` XML element.
"""


def make_citation_information(info):
    # type: (Citation_Information) -> NodeBuilder
    """
    Return a placeholder ``<Citation_Information />`` XML element.
    """
    return _citation_information({
        'description': info.description,
        'publication_year': info.publication_year
    })
