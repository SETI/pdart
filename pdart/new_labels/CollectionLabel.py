"""
Functionality to build a collection label using a SQLite database.
"""

from typing import TYPE_CHECKING

from pdart.new_labels.CitationInformation \
    import make_placeholder_citation_information
from pdart.new_labels.CollectionLabelXml import make_label
from pdart.new_labels.Utils import lidvid_to_lid, lidvid_to_vid
from pdart.xml.Pretty import pretty_and_verify

if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


# TODO Should probably test document_collection independently.


def make_collection_label(bundle_db, collection_lidvid, verify):
    # type: (BundleDB, str, bool) -> unicode
    """
    Create the label text for the collection having this LIDVID using
    the bundle database.  If verify is True, verify the label against
    its XML and Schematron schemas.  Raise an exception if either
    fails.
    """
    collection_lid = lidvid_to_lid(collection_lidvid)
    collection_vid = lidvid_to_vid(collection_lidvid)
    collection = bundle_db.get_collection(collection_lidvid)
    suffix = collection.suffix
    proposal_id = bundle_db.get_bundle().proposal_id
    try:
        inventory_name = 'collection_%s.csv' % collection.prefix
    except:
        # Document collections won't have prefixes.
        inventory_name = 'collection.csv'

    label = make_label({
        'collection_lid': collection_lid,
        'collection_vid': collection_vid,
        'suffix': suffix,
        'proposal_id': str(proposal_id),
        'Citation_Information': make_placeholder_citation_information(
            collection_lid),
        'inventory_name': inventory_name
    }).toxml()

    return pretty_and_verify(label, verify)
