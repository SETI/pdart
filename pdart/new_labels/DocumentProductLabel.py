from contextlib import closing
from datetime import date
import sys

from pdart.new_labels.DocumentProductLabelXml import *
from pdart.new_labels.Utils import lidvid_to_lid
from pdart.xml.Pretty import pretty_and_verify

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


def make_document_product_label(bundle_db, document_product_lidvid, verify):
    # type: (BundleDB, str, bool) -> unicode
    """
    Create the label text for the document product in the bundle
    having this :class:`~pdart.pds4.LID` using the database
    connection.  If verify is True, verify the label against its XML
    and Schematron schemas.  Raise an exception if either fails.
    """
    # TODO Should we be using LIDVIDs instead of LIDs?
    bundle = bundle_db.get_bundle()
    bundle_lid = lidvid_to_lid(bundle.lidvid)
    proposal_id = bundle.proposal_id
    title = 'Summary of the observation plan for HST proposal %d' % proposal_id

    product_lid = lidvid_to_lid(document_product_lidvid)

    label = make_label({
            'bundle_lid': bundle_lid,
            'product_lid': product_lid,
            'title': title,
            'publication_date': date.today().isoformat(),
            'Citation_Information': make_citation_information(bundle_lid,
                                                              proposal_id),
            'Document_Edition': make_document_edition(
                '0.0',
                [('phase2.txt', '7-Bit ASCII Text')])
            }).toxml()
    return pretty_and_verify(label, verify)


#     product_lid = bundle_lid + ':document:phase2'
#     with closing(conn.cursor()) as cursor:
#         # get some unknown info
#         (label_fp, proposal_id) = get_document_product_info(cursor,
#                                                             product_lid)
#         pass
#
#     title = 'Summary of the observation plan for HST proposal %d' % \
# proposal_id
#
#     label = make_label({
#             'bundle_lid': bundle_lid,
#             'product_lid': product_lid,
#             'title': title,
#             'publication_date': date.today().isoformat(),
#             'Citation_Information': make_citation_information(bundle_lid,
#                                                               proposal_id),
#             'Document_Edition': make_document_edition(
#                 '0.0',
#                 [('phase2.txt', '7-Bit ASCII Text')])
#             }).toxml()
#     label = pretty_print(label)
#
#     with open(label_fp, 'w') as f:
#         f.write(label)
#
#     print 'product label for', product_lid
#     sys.stdout.flush()
#
#     if verify:
#         verify_label_or_raise(label)
#
#     return label
