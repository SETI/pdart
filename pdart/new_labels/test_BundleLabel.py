import unittest

from pdart.new_db.BundleDB import create_bundle_db_in_memory
from pdart.new_labels.BundleLabel import *
from pdart.new_labels.Utils import golden_file_contents

_BUNDLE_LIDVID = 'urn:nasa:pds:hst_09059::1.3'
_COLLECTION_LIDVID = 'urn:nasa:pds:hst_09059:data_acs_raw::1.2'
_FITS_PRODUCT_LIDVID = 'urn:nasa:pds:hst_09059:data_acs_raw:j6gp01lzq_raw::1.2'


class Test_BundleLabel(unittest.TestCase):
    def setUp(self):
        # type: () -> None
        self.db = create_bundle_db_in_memory()
        self.db.create_tables()
        self.db.create_bundle(_BUNDLE_LIDVID)
        self.db.create_non_document_collection(_COLLECTION_LIDVID,
                                               _BUNDLE_LIDVID)

        self.db.create_fits_product(_FITS_PRODUCT_LIDVID, _COLLECTION_LIDVID)

    def test_make_bundle_label(self):
        # type: () -> None
        label = make_bundle_label(self.db, True)
        expected_label = golden_file_contents('test_BundleLabel.golden.xml')
        self.assertEqual(expected_label, label)
