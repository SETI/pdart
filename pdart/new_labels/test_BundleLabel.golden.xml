<?xml version="1.0"?>
<?xml-model href="https://pds.nasa.gov/pds4/pds/v1/PDS4_PDS_1A10.sch" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<Product_Bundle xmlns="http://pds.nasa.gov/pds4/pds/v1" xmlns:pds="http://pds.nasa.gov/pds4/pds/v1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://pds.nasa.gov/pds4/pds/v1 https://pds.nasa.gov/pds4/pds/v1/PDS4_PDS_1A10.xsd">
  <Identification_Area>
    <logical_identifier>urn:nasa:pds:hst_09059</logical_identifier>
    <version_id>1.3</version_id>
    <title>This bundle contains images obtained from HST Observing Program
9059.</title>
    <information_model_version>1.10.1.0</information_model_version>
    <product_class>Product_Bundle</product_class>
    <Citation_Information>
      <publication_year>2001</publication_year>
      <description>&lt;description&gt;</description>
    </Citation_Information>
  </Identification_Area>
  <Bundle>
    <bundle_type>Archive</bundle_type>
  </Bundle>
  <Bundle_Member_Entry>
    <lidvid_reference>urn:nasa:pds:hst_09059:data_acs_raw::1.2</lidvid_reference>
    <member_status>Primary</member_status>
    <reference_type>bundle_has_data_collection</reference_type>
  </Bundle_Member_Entry>
</Product_Bundle>
